### PR TITLE
CAFF-19 [incomplete]: Load and run CV model

### PR DESCRIPTION
**Pull request by kajananchinniah**
Doesn't close CAFF-19 yet since we need to convert the segmentation mask into a usable form for virtual layers.

Adds:
- Ability to load onnx model via launch file
- Node to take an image on `/image` and then publish a binary segmentation mask on `/lane_detection_mask` 
- Preliminary instructions to install onnxruntime

Test:
- Publish an image on `/image`
- Compile and build `catkin_make && source devel/setup.bash`
- Launch the node: `roslaunch lane_detection lane_detection.launch`. Make sure to update the filepath to a valid onnx model
- Launch rviz and visualize `/lane_detection_mask`

Todo:
- Need to convert the segmentation mask into whatever virtual layer needs. @1999michael to define what he needs.

Credits to @spencerteetaert for doing most of the code, I just refactored & made the ros wrapper.

We should do the conversion in a separate PR since it's not defined yet.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -8,3 +8,6 @@
 
 # VSCode
 .vscode
+
+# CV models (too big) 
+*.onnx
```
Comment (kajananchinniah) [inline .gitignore:13]: Need to fix end of line

File: README.md [hunk 2]
```diff
@@ -69,6 +69,23 @@ sudo apt-get install ros-melodic-hector-gazebo-plugins
 Custom built world(s) representing the IGVC competition can be found in the [`/worlds`](./worlds) package. To install them for use in the Gazebo simulator, run the `./install_models.sh` script found in the `/worlds/models` folder.
 > **NOTE:** The install script copies specific contents of `/worlds/models` to `~/.gazebo/models`
 
+### Install ONNXRuntime ###
```
Comment (kajananchinniah) [inline README.md:72]: @spencerteetaert could you write the instructions for this part...

File: cv/.gitignore [hunk 3]
```diff
@@ -0,0 +1,51 @@
+devel/
```
Comment (kajananchinniah) [inline cv/.gitignore:1]: @spencerteetaert 

Was this change necessary? We can move it to the root if you think it is

File: cv/.gitignore [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+/planning/docs
+/planning/src
+
+*~
+
+# Emacs
+.#*
+
+# Catkin custom files
+CATKIN_IGNORE
```
Comment (kajananchinniah) [inline cv/.gitignore:51]: Fix end of line

File: cv/lane_detection/launch/lane_detection.launch [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<launch>
+    <node
+        name="cv_model"
+        pkg="lane_detection"
+        type="lane_detection_inference_node"
+        output="screen"
+    >
+        <param name="instance_name" value="unet_inference"/>
+        <param name="model_path" value="/tmp/unet_with_sigmoid.onnx"/>
```
Comment (kajananchinniah) [inline cv/lane_detection/launch/lane_detection.launch:9]: @spencerteetaert 

Where to put model path?

File: cv/lane_detection/package.xml [hunk 6]
```diff
@@ -0,0 +1,61 @@
+<?xml version="1.0"?>
+<package format="2">
+  <name>lane_detection</name>
+  <version>0.0.0</version>
+  <description>The lane_detection package</description>
+
+  <!-- One maintainer tag required, multiple allowed, one person per tag -->
+  <!-- Example:  -->
+  <!-- <maintainer email="jane.doe@example.com">Jane Doe</maintainer> -->
+  <maintainer email="kajanan@todo.todo">kajanan</maintainer>
```
Comment (kajananchinniah) [inline cv/lane_detection/package.xml:10]: Add spencer as a maintainer

File: cv/lane_detection/package.xml [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+
+  <!-- One license tag required, multiple allowed, one license per tag -->
+  <!-- Commonly used license strings: -->
+  <!--   BSD, MIT, Boost Software License, GPLv2, GPLv3, LGPLv2.1, LGPLv3 -->
+  <license>TODO</license>
+
+
+  <!-- Url tags are optional, but multiple are allowed, one per tag -->
```
Comment (kajananchinniah) [inline cv/lane_detection/package.xml:19]: Clean this up / delete most of the comments

File: cv/lane_detection/src/lane_detection_inference_node.cpp [hunk 8]
```diff
@@ -0,0 +1,48 @@
+#include <boost/bind/placeholders.hpp>
+#include <string>
+
+#include "cv_bridge/cv_bridge.h"
+#include "lane_detection/lane_detection_inference_node.h"
+
+LaneDetectionInferenceNode::LaneDetectionInferenceNode(ros::NodeHandle nh)
+    : nh_{nh}, it_{nh_}, inferencer_{[&]() {
```
Comment (kajananchinniah) [inline cv/lane_detection/src/lane_detection_inference_node.cpp:8]: Fix formatting here

File: cv/lane_detection/src/lane_detection_inferencer.cpp [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+#include "onnxruntime_cxx_api.h"
+#include "opencv2/opencv.hpp"
+
+namespace
+{
+
+constexpr int NETWORK_INPUT_WIDTH{330};
+constexpr int NETWORK_INPUT_HEIGHT{180};
+constexpr int NETWORK_INPUT_CHANNEL{5};
+constexpr int NETWORK_OUTPUT_WIDTH{330};
```
Comment (kajananchinniah) [inline cv/lane_detection/src/lane_detection_inferencer.cpp:21]: We probably shouldn't have a separate constant for input and output in hindsight (since we're doing segmentation here)

File: cv/lane_detection/src/lane_detection_inferencer.cpp [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+    return std::accumulate(v.begin(), v.end(), 1, std::multiplies<T>());
+}
+
+uint8_t getMedian(std::vector<uint8_t> input)
+{
+    if (input.empty()) {
+        throw std::invalid_argument("Received empty input vector!");
+    }
+
+    std::nth_element(input.begin(), input.begin() + input.size() / 2,
```
Comment (kajananchinniah) [inline cv/lane_detection/src/lane_detection_inferencer.cpp:35]: @spencerteetaert 

Is this valid for even sized inputs ?

File: cv/lane_detection/src/lane_detection_inferencer.cpp [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+        cv::Mat(quad, CV_8UC1)
+            .reshape(1, static_cast<int>(grayscaled_image.rows / 2))};
+    cv::Mat edges;
+    cv::Canny(quad_cv_mat, edges, lower_bound, upper_bound);
+    return edges;
+}
+
+cv::Mat getEdgeChannel(cv::Mat grayscaled_image)
+{
+    cv::Mat quad1_edges{getQuadrantEdges(
```
Comment (kajananchinniah) [inline cv/lane_detection/src/lane_detection_inferencer.cpp:73]: This part still isn't really clean, but I personally think it's good enough

Thoughts?

File: cv/lane_detection/src/lane_detection_inferencer.cpp [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+
+    // stitch edges together
+    const int new_rows{std::max(quad1_edges.rows + quad3_edges.rows,
+                                quad2_edges.rows + quad4_edges.rows)};
+    const int new_cols{std::max(quad1_edges.cols + quad3_edges.cols,
+                                quad2_edges.cols + quad4_edges.cols)};
+
+    cv::Mat edges_mask = cv::Mat::zeros(new_rows, new_cols, CV_8UC1);
+
+    quad1_edges.copyTo(
```
Comment (kajananchinniah) [inline cv/lane_detection/src/lane_detection_inferencer.cpp:95]: Copies here might not be necessary @spencerteetaert 

I didn't bother trying to optimize but do you think we should try to avoid it ?

File: cv/lane_detection/src/lane_detection_inferencer.cpp [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+}
+
+cv::Mat getInverseEdgeChannel(cv::Mat edge_channel)
+{
+    cv::Mat inverse_edge_channel;
+    cv::bitwise_not(edge_channel, inverse_edge_channel);
+    return inverse_edge_channel;
+}
+
+cv::Mat combineImages(cv::Mat normalized_img, cv::Mat normalized_edges,
```
Comment (kajananchinniah) [inline cv/lane_detection/src/lane_detection_inferencer.cpp:114]: In hindsight we should be able to combine this function and the `convertHWC2CHW` function

We split the channels, then merge them back together lthen split them again and then concatenate them I think lol

@spencerteetaert I don't rmr why we did this way anymore, was there a reason ?

File: cv/lane_detection/src/lane_detection_inferencer.cpp [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+    cv::Mat normalized_img;
+    resized_img.convertTo(normalized_img, CV_32F, 1.0f / 255.0f, 0.0f);
+
+    return convertHWC2CHW(
+        combineImages(normalized_img, normalized_edges, normalized_edges_inv));
+}
+
+cv::Mat LaneDetectionInferencer::performInference(cv::Mat preprocessed_img)
+{
+    std::vector<float> input_tensor_values(input_tensor_size_);
```
Comment (kajananchinniah) [inline cv/lane_detection/src/lane_detection_inferencer.cpp:213]: We can probably cache `input_tensor_values` instead of creating on each function call

File: cv/lane_detection/src/lane_detection_inferencer.cpp [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+    output_tensors.push_back(Ort::Value::CreateTensor<float>(
+        memory_info_, output_tensor_values_.data(), output_tensor_size_,
+        output_dims_.data(), output_dims_.size()));
+
+    session_.Run(Ort::RunOptions{nullptr}, input_names_.data(),
+                 input_tensors.data(), 1, output_names_.data(),
+                 output_tensors.data(), 1);
+
+    // TODO: do not use C style cast, look for safer way to do this
+    return cv::Mat(NETWORK_OUTPUT_HEIGHT, NETWORK_OUTPUT_WIDTH, CV_32FC1,
```
Comment (kajananchinniah) [inline cv/lane_detection/src/lane_detection_inferencer.cpp:233]: @spencerteetaert 

Do you know what the underlying data is? I would prefer avoiding a cast if possible but I'm fine with it since everything works I guess

**Review by kajananchinniah — COMMENTED** [review]
I left initial review comments based on a quick skim of the code and tagged spencer where relevant. I can't actually review it since I'm an assignee

My comments serve mostly as a reminder for things I should get to, but feel free to reply to the comments I left

File: cv/lane_detection/src/lane_detection_inference_node.cpp [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+          nh_.param<std::string>("instance_name", instance_name_, "");
+          ROS_INFO("Instance name: %s", instance_name_.c_str());
+          nh_.param<std::string>("model_path", model_path_, "");
+          ROS_INFO("Model path: %s", model_path_.c_str());
+          return LaneDetectionInferencer(instance_name_, model_path_);
```
Comment (kajananchinniah) [inline cv/lane_detection/src/lane_detection_inference_node.cpp:13]: Don't need to actually store the instance or model path

Comment (kajananchinniah) [reply cv/lane_detection/src/lane_detection_inference_node.cpp:13]: Doesn't hurt to store anyways

File: cv/lane_detection/launch/lane_detection.launch (hunk 5 above)
Comment (spencerteetaert) [reply cv/lane_detection/launch/lane_detection.launch:9]: I hadn't figured out relative paths in ros environments so had it as a global on my system. Ideally if we can have the models in a folder like /lane_detection/models/model.onnx that would be prefered. Something that's easy to switch out. We can set a global path on the jetson/laptop systems as well, but it may be okay to just leave it as a placeholder on the git cause we aren't wanting to upload the .onnx model

File: cv/lane_detection/src/lane_detection_inferencer.cpp (hunk 12 above)
Comment (spencerteetaert) [reply cv/lane_detection/src/lane_detection_inferencer.cpp:95]: there definitely is room for optimization. I'm not too too concerned with it already running at 40+fps. We can spend more time on it if it's under performing with full system running

File: cv/lane_detection/src/lane_detection_inferencer.cpp (hunk 13 above)
Comment (spencerteetaert) [reply cv/lane_detection/src/lane_detection_inferencer.cpp:114]: Ease of implementation. The convert HWC2CHW function provided the numpy .permute equivelant for a 3D image and we never looked at merging it with the other concat functions we were doing. Good catch, probably most of these matrix copying/concatenating operations can be reduced.

File: cv/lane_detection/src/lane_detection_inferencer.cpp (hunk 15 above)
Comment (spencerteetaert) [reply cv/lane_detection/src/lane_detection_inferencer.cpp:233]: As far as I can tell it's some Ort value class. Not sure what the actual datatype is, I would assume in this case it's a wrapper for array of float32 by default

File: cv/lane_detection/launch/lane_detection.launch (hunk 5 above)
Comment (kajananchinniah) [reply cv/lane_detection/launch/lane_detection.launch:9]: We'll deal w/ it later then

File: cv/lane_detection/src/lane_detection_inferencer.cpp (hunk 12 above)
Comment (kajananchinniah) [reply cv/lane_detection/src/lane_detection_inferencer.cpp:95]: Agreed

File: cv/lane_detection/src/lane_detection_inferencer.cpp (hunk 15 above)
Comment (kajananchinniah) [reply cv/lane_detection/src/lane_detection_inferencer.cpp:233]: I'm fine with the cast then, we'll worry about fixing it later if it's problematic

File: cv/lane_detection/src/lane_detection_inferencer.cpp (hunk 13 above)
Comment (kajananchinniah) [reply cv/lane_detection/src/lane_detection_inferencer.cpp:114]: We can optimize later, as you mentioned in a separate comment, performance was good enough

File: cv/lane_detection/src/lane_detection_inferencer.cpp (hunk 10 above)
Comment (kajananchinniah) [reply cv/lane_detection/src/lane_detection_inferencer.cpp:35]: Won't work for even sized inputs, but it's probably fine

File: README.md (hunk 2 above)
Comment (kajananchinniah) [reply README.md:72]: I added instructions, but we need to verify them later

**Review by spencerteetaert — APPROVED** [review]
All looks good. Keeping branch open for addition of post-processing script later.
